src/k8s/pkg/client/dqlite/remove.go

```diff
@@ -27,21 +31,48 @@ func (c *Client) RemoveNodeByAddress(ctx context.Context, address string) error
 
 		case member.Address != address && member.Role == Voter:
 			clusterHasOtherVoters = true
+
+		case member.Address != address && member.Role == Spare:
+			// This is only used in a two node setup, where the leader node is removed.
+			// The free spare node will be promoted to leader before the existing leader is removed.
+			freeSpareNode = &member
 		}
 	}
 
 	if !memberExists {
 		return fmt.Errorf("cluster does not have a node with address %v", address)
 	}
 
-	// TODO: consider using client.Transfer() for a different node to become leader
 	if !clusterHasOtherVoters {
-		return fmt.Errorf("not removing node because there are no other voter members")
-	}
+		if freeSpareNode == nil {
+			// This normally should not happen. There should always be a backup node, except
+			// if one tries to remove the last node in the cluster.
+			return fmt.Errorf("cannot transfer dqlite leadership as there is no remaining spare node")
+		}
 
-	if err := client.Remove(ctx, memberToRemove.ID); err != nil {
-		return fmt.Errorf("failed to remove node %#v from dqlite cluster: %w", memberToRemove, err)
+		// Leadership can only be transfered to a voter or standby node.
+		// Therefore the remaining node in the cluster needs to be promoted first.
+		if err := client.Assign(ctx, freeSpareNode.ID, Voter); err != nil {
+			return fmt.Errorf("failed to assign voter role to %d: %w", freeSpareNode.ID, err)
+		}
+		// Transfer leadership to remaining node in cluster.
+		if err := client.Transfer(ctx, freeSpareNode.ID); err != nil {
+			return fmt.Errorf("failed to transfer leadership to %d: %w", freeSpareNode.ID, err)
+		}
+		// Recreate client to point to the new leader.
+		client, err = c.clientGetter(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to create dqlite client: %w", err)
+		}
 	}
 
-	return nil
+	// Remove the node from the cluster. Retry as the leadership transfer might still be in progress.
+	// For a large database this might take some time.
+	return control.RetryFor(ctx, 10, func() error {
+		if err := client.Remove(ctx, memberToRemove.ID); err != nil {
+			time.Sleep(5 * time.Second)
```

Comment:
this can be dangerous. is this backoff? if so, it should be specified properly in `control.RetryFor`, and `control.RetryFor` should use `<-time.After(timeout)` instead. For some more background consider the difference between:

```go
select {
case <-ctx.Done():
  // do stuff
case <-time.After(5 * time.Second):
}
```

and 

```go
time.Sleep(5 * time.Second)
<-ctx.Done()

// do stuff
```

in the second case, the `time.Sleep` takes control away from our code, and a context cancellation will block until the sleep completes.

Reply:
Good point, moved the delay to the `RetryFor` function.